tests/testsuite/profile_targets.rs

```diff
@@ -0,0 +1,703 @@
+            name = "bdep"
+            version = "0.0.1"
+
+            [dependencies]
+            bar = { path = "../bar" }
+        "#)
+        .file("bdep/src/lib.rs", "extern crate bar;")
+        .build()
+}
+
+#[test]
+fn profile_selection_build() {
+    let p = all_target_project();
+
+    // Build default targets.
+    // NOTES:
+    // - bdep `panic` is not set because it thinks `build.rs` is a plugin.
+    // - bar `panic` is not set because it is shared with `bdep`.
+    // - build_script_build is built without panic because it thinks `build.rs` is a plugin.
+    assert_that(p.cargo("build -vv"), execs().with_status(0).with_stderr_unordered("\
+[COMPILING] bar [..]
+[RUNNING] `rustc --crate-name bar bar[/]src[/]lib.rs --crate-type lib --emit=dep-info,link -C codegen-units=1 -C debuginfo=2 [..]
+[COMPILING] bdep [..]
+[RUNNING] `rustc --crate-name bdep bdep[/]src[/]lib.rs --crate-type lib --emit=dep-info,link -C codegen-units=1 -C debuginfo=2 [..]
+[COMPILING] foo [..]
+[RUNNING] `rustc --crate-name build_script_build build.rs --crate-type bin --emit=dep-info,link -C codegen-units=1 -C debuginfo=2 [..]
+[RUNNING] `[..][/]target[/]debug[/]build[/]foo-[..][/]build-script-build`
+foo custom build PROFILE=debug DEBUG=true OPT_LEVEL=0
+[RUNNING] `rustc --crate-name foo src[/]lib.rs --crate-type lib --emit=dep-info,link -C panic=abort -C codegen-units=1 -C debuginfo=2 [..]
+[RUNNING] `rustc --crate-name foo src[/]main.rs --crate-type bin --emit=dep-info,link -C panic=abort -C codegen-units=1 -C debuginfo=2 [..]
+[FINISHED] dev [unoptimized + debuginfo] [..]
+"));
+    assert_that(
+        p.cargo("build -vv"),
+        execs().with_status(0).with_stderr_unordered(
+            "\
+[FRESH] bar [..]
+[FRESH] bdep [..]
+[FRESH] foo [..]
+[FINISHED] dev [unoptimized + debuginfo] [..]
+",
+        ),
+    );
+}
+
+#[test]
+fn profile_selection_build_release() {
+    let p = all_target_project();
+
+    // Build default targets, release.
+    assert_that(p.cargo("build --release -vv"),
+        execs().with_status(0).with_stderr_unordered("\
+[COMPILING] bar [..]
+[RUNNING] `rustc --crate-name bar bar[/]src[/]lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C codegen-units=2 [..]
+[COMPILING] bdep [..]
+[RUNNING] `rustc --crate-name bdep bdep[/]src[/]lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C codegen-units=2 [..]
+[COMPILING] foo [..]
+[RUNNING] `rustc --crate-name build_script_build build.rs --crate-type bin --emit=dep-info,link -C opt-level=3 -C codegen-units=2 [..]
+[RUNNING] `[..][/]target[/]release[/]build[/]foo-[..][/]build-script-build`
+foo custom build PROFILE=release DEBUG=false OPT_LEVEL=3
+[RUNNING] `rustc --crate-name foo src[/]lib.rs --crate-type lib --emit=dep-info,link -C opt-level=3 -C panic=abort -C codegen-units=2 [..]
+[RUNNING] `rustc --crate-name foo src[/]main.rs --crate-type bin --emit=dep-info,link -C opt-level=3 -C panic=abort -C codegen-units=2 [..]
+[FINISHED] release [optimized] [..]
```

Comment:
I guess I don't understand something about panic handling... How this even works? We link `bar`, which is compiled with `panic=unwind` into `foo`, which is compiled with `panic=abort` and `rustc` does not complain. Is it the expected behavior? Testing locally, I see that `unwind -> abort` direction works, and `abort -> unwind` fails. Does this mean that it is always safe to compile dependencies with `panic = unwind`? 


This probably has nothing to do with the PR though, because the behavior has not changed here.

Reply:
Hm, yea, I don't know much how panic handling is implemented.  I did a test where I attempt to panic in a dependency that was compiled with unwind, and it still aborts.  It looks like it just affects whether `libpanic_abort` or `libpanic_unwind` is linked.  I'm guessing since the API is the same, the dependencies don't know the difference.

Reply:
>It looks like it just affects whether libpanic_abort or libpanic_unwind is linked. I

I would guess that it also affects generation of landing pads? So we might be getting a bunch of dead code here?

@alexcrichton could you explain how mixing panic modes is supposed to work?

Reply:
Ah yes the observed behavior here is expected, we , for example, ship panic=unwind libstd to everyone but still allow compiling with panic=abort